Dataset location
https://www.kaggle.com/datasets/mrmorj/dataset-of-songs-in-spotify

Business case possibilities:
* song recommendation based on similar songs?

# Setup

Import necessary libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import tensorflow as tf

In [ ]:
data = pd.read_csv("genres_v2.csv")
data.head()

/opt/conda/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3378: DtypeWarning: Columns (19) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.831   0.814    2    -7.364     1       0.4200        0.0598   
1         0.719   0.493    8    -7.230     1       0.0794        0.4010   
2         0.850   0.893    5    -4.783     1       0.0623        0.0138   
3         0.476   0.781    0    -4.710     1       0.1030        0.0237   
4         0.798   0.624    2    -7.668     1       0.2930        0.2170   

   instrumentalness  liveness  valence  ...                      id  \
0          0.013400    0.0556   0.3890  ...  2Vc6NJ9PW9gD9q343XFRKx   
1          0.000000    0.1180   0.1240  ...  7pgJBLVz5VmnL7uGHmRj6p   
2          0.000004    0.3720   0.0391  ...  0vSWgAlfpye0WCGeNmuNhy   
3          0.000000    0.1140   0.1750  ...  0VSXnJqQkwuH2ei1nOQ1nu   
4          0.000000    0.1660   0.5910  ...  4jCeguq9rMTlbMmPHuO7S3   

                                    uri  \
0  spotify:track:2Vc6NJ9PW9gD9q343XFRKx   
1  spotify:track:7pgJBLVz5VmnL7uGHmRj6p   
2  spotify:track:0vSWgAlfpye0WCGeNmuNhy   
3  spotify:track:0VSXnJqQkwuH2ei1nOQ1nu   
4  spotify:track:4jCeguq9rMTlbMmPHuO7S3   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/2Vc6NJ9PW9gD...   
1  https://api.spotify.com/v1/tracks/7pgJBLVz5Vmn...   
2  https://api.spotify.com/v1/tracks/0vSWgAlfpye0...   
3  https://api.spotify.com/v1/tracks/0VSXnJqQkwuH...   
4  https://api.spotify.com/v1/tracks/4jCeguq9rMTl...   

                                        analysis_url duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/2Vc6...      124539   
1  https://api.spotify.com/v1/audio-analysis/7pgJ...      224427   
2  https://api.spotify.com/v1/audio-analysis/0vSW...       98821   
3  https://api.spotify.com/v1/audio-analysis/0VSX...      123661   
4  https://api.spotify.com/v1/audio-analysis/4jCe...      123298   

  time_signature      genre                                      song_name  \
0              4  Dark Trap                            Mercury: Retrograde   
1              4  Dark Trap                                      Pathology   
2              4  Dark Trap                                       Symbiote   
3              3  Dark Trap  ProductOfDrugs (Prod. The Virus and Antidote)   
4              4  Dark Trap                                          Venom   

  Unnamed: 0 title  
0        NaN   NaN  
1        NaN   NaN  
2        NaN   NaN  
3        NaN   NaN  
4        NaN   NaN  

[5 rows x 22 columns]

# EDA

In [2]:
data.info()

NameError: ignored

In [ ]:
# sns.pairplot(data, diag_kind = 'kde')
# plt.show()

**Danceability**

In [ ]:
data["danceability"].value_counts(dropna=False).sort_index()

0.0651    2
0.0753    1
0.0891    2
0.0979    1
0.1060    2
         ..
0.9820    1
0.9830    2
0.9840    1
0.9850    4
0.9880    3
Name: danceability, Length: 890, dtype: int64

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["danceability"])
# plt.show()

**Energy**

In [ ]:
# data["energy"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["energy"])
# plt.show()

**Key**

In [ ]:
# data["key"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["key"])
# plt.show()

**Loudness**

In [ ]:
# data["loudness"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["loudness"])
# plt.show()

**Mode**

In [ ]:
# data["mode"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["mode"])
# plt.show()

**Speechiness**

In [ ]:
# data["speechiness"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["speechiness"])
# plt.show()

**Acousticness**

In [ ]:
# data["acousticness"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["acousticness"])
# plt.show()

**Instrumentalness**

In [ ]:
# data["instrumentalness"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["instrumentalness"])
# plt.show()

**Liveness**

In [ ]:
# data["liveness"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["liveness"])
# plt.show()

**Valence**

In [ ]:
# data["valence"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["energy"])
# plt.show()

**Tempo**

In [ ]:
# data["tempo"].value_counts(dropna=False).sort_index()

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["energy"])
# plt.show()

**Type**

In [ ]:
# data["type"].value_counts(dropna=False).sort_index()

In [ ]:
# data["type"].unique()

**ID**

In [ ]:
data["id"].value_counts(dropna=False).sort_index()

001dY1wrXF0Vfo40Kkqv4R    2
001rJKlws3dVKuURJQdnIk    1
00349h1hy2raWiakxSDfsU    1
003FTlCpBTM4eSqYSWPv4H    1
003VDDA7J3Xb2ZFlNx7nIZ    1
                         ..
7zrH5Yxm0GYeQAKyy4ctp5    1
7zs4U1U5Rbj5k7nQ1vuuGF    1
7zw5QFPrerWf8XzZW70z3L    1
7zxtj0SKcfBMYWnf7nYrsG    1
7zyzxDwgCam5DYP3ddWMjw    2
Name: id, Length: 35877, dtype: int64

In [ ]:
data["id"].value_counts(dropna=False, ascending=False)

7D5qMGajdfznFcXFArk5VE    9
5cRDn5aGMLvWsldoRmOOz0    8
5lXY6PTuWXOludKy4zDQwM    8
3BSuC9Y6MhalaLlhyNbTYZ    8
2KVv2zAjDHxymySY5EIaAz    7
                         ..
6hEHxpZpvZwt1v4NoGwjKr    1
5TvdhNXoHbeGeWtjCYhFuy    1
0PIhZHuUMkxYIueyOJ9rSh    1
2V28RKTOabMKk2xvIDyCEz    1
6MAAMZImxcvYhRnxDLTufD    1
Name: id, Length: 35877, dtype: int64

**Uri**

In [ ]:
data["uri"].value_counts(dropna=False).sort_index()

spotify:track:001dY1wrXF0Vfo40Kkqv4R    2
spotify:track:001rJKlws3dVKuURJQdnIk    1
spotify:track:00349h1hy2raWiakxSDfsU    1
spotify:track:003FTlCpBTM4eSqYSWPv4H    1
spotify:track:003VDDA7J3Xb2ZFlNx7nIZ    1
                                       ..
spotify:track:7zrH5Yxm0GYeQAKyy4ctp5    1
spotify:track:7zs4U1U5Rbj5k7nQ1vuuGF    1
spotify:track:7zw5QFPrerWf8XzZW70z3L    1
spotify:track:7zxtj0SKcfBMYWnf7nYrsG    1
spotify:track:7zyzxDwgCam5DYP3ddWMjw    2
Name: uri, Length: 35877, dtype: int64

In [ ]:
data["uri"].value_counts(dropna=False, ascending=False)

spotify:track:7D5qMGajdfznFcXFArk5VE    9
spotify:track:5cRDn5aGMLvWsldoRmOOz0    8
spotify:track:5lXY6PTuWXOludKy4zDQwM    8
spotify:track:3BSuC9Y6MhalaLlhyNbTYZ    8
spotify:track:2KVv2zAjDHxymySY5EIaAz    7
                                       ..
spotify:track:6hEHxpZpvZwt1v4NoGwjKr    1
spotify:track:5TvdhNXoHbeGeWtjCYhFuy    1
spotify:track:0PIhZHuUMkxYIueyOJ9rSh    1
spotify:track:2V28RKTOabMKk2xvIDyCEz    1
spotify:track:6MAAMZImxcvYhRnxDLTufD    1
Name: uri, Length: 35877, dtype: int64

**track_href**

In [ ]:
data["track_href"].value_counts(dropna=False).sort_index()

https://api.spotify.com/v1/tracks/001dY1wrXF0Vfo40Kkqv4R    2
https://api.spotify.com/v1/tracks/001rJKlws3dVKuURJQdnIk    1
https://api.spotify.com/v1/tracks/00349h1hy2raWiakxSDfsU    1
https://api.spotify.com/v1/tracks/003FTlCpBTM4eSqYSWPv4H    1
https://api.spotify.com/v1/tracks/003VDDA7J3Xb2ZFlNx7nIZ    1
                                                           ..
https://api.spotify.com/v1/tracks/7zrH5Yxm0GYeQAKyy4ctp5    1
https://api.spotify.com/v1/tracks/7zs4U1U5Rbj5k7nQ1vuuGF    1
https://api.spotify.com/v1/tracks/7zw5QFPrerWf8XzZW70z3L    1
https://api.spotify.com/v1/tracks/7zxtj0SKcfBMYWnf7nYrsG    1
https://api.spotify.com/v1/tracks/7zyzxDwgCam5DYP3ddWMjw    2
Name: track_href, Length: 35877, dtype: int64

In [ ]:
data["track_href"].value_counts(dropna=False, ascending=False)

https://api.spotify.com/v1/tracks/7D5qMGajdfznFcXFArk5VE    9
https://api.spotify.com/v1/tracks/5cRDn5aGMLvWsldoRmOOz0    8
https://api.spotify.com/v1/tracks/5lXY6PTuWXOludKy4zDQwM    8
https://api.spotify.com/v1/tracks/3BSuC9Y6MhalaLlhyNbTYZ    8
https://api.spotify.com/v1/tracks/2KVv2zAjDHxymySY5EIaAz    7
                                                           ..
https://api.spotify.com/v1/tracks/6hEHxpZpvZwt1v4NoGwjKr    1
https://api.spotify.com/v1/tracks/5TvdhNXoHbeGeWtjCYhFuy    1
https://api.spotify.com/v1/tracks/0PIhZHuUMkxYIueyOJ9rSh    1
https://api.spotify.com/v1/tracks/2V28RKTOabMKk2xvIDyCEz    1
https://api.spotify.com/v1/tracks/6MAAMZImxcvYhRnxDLTufD    1
Name: track_href, Length: 35877, dtype: int64

**analysis_url**

In [ ]:
data["analysis_url"].value_counts(dropna=False).sort_index()

https://api.spotify.com/v1/audio-analysis/001dY1wrXF0Vfo40Kkqv4R    2
https://api.spotify.com/v1/audio-analysis/001rJKlws3dVKuURJQdnIk    1
https://api.spotify.com/v1/audio-analysis/00349h1hy2raWiakxSDfsU    1
https://api.spotify.com/v1/audio-analysis/003FTlCpBTM4eSqYSWPv4H    1
https://api.spotify.com/v1/audio-analysis/003VDDA7J3Xb2ZFlNx7nIZ    1
                                                                   ..
https://api.spotify.com/v1/audio-analysis/7zrH5Yxm0GYeQAKyy4ctp5    1
https://api.spotify.com/v1/audio-analysis/7zs4U1U5Rbj5k7nQ1vuuGF    1
https://api.spotify.com/v1/audio-analysis/7zw5QFPrerWf8XzZW70z3L    1
https://api.spotify.com/v1/audio-analysis/7zxtj0SKcfBMYWnf7nYrsG    1
https://api.spotify.com/v1/audio-analysis/7zyzxDwgCam5DYP3ddWMjw    2
Name: analysis_url, Length: 35877, dtype: int64

In [ ]:
data["analysis_url"].value_counts(dropna=False, ascending=False)

https://api.spotify.com/v1/audio-analysis/7D5qMGajdfznFcXFArk5VE    9
https://api.spotify.com/v1/audio-analysis/5cRDn5aGMLvWsldoRmOOz0    8
https://api.spotify.com/v1/audio-analysis/5lXY6PTuWXOludKy4zDQwM    8
https://api.spotify.com/v1/audio-analysis/3BSuC9Y6MhalaLlhyNbTYZ    8
https://api.spotify.com/v1/audio-analysis/2KVv2zAjDHxymySY5EIaAz    7
                                                                   ..
https://api.spotify.com/v1/audio-analysis/6hEHxpZpvZwt1v4NoGwjKr    1
https://api.spotify.com/v1/audio-analysis/5TvdhNXoHbeGeWtjCYhFuy    1
https://api.spotify.com/v1/audio-analysis/0PIhZHuUMkxYIueyOJ9rSh    1
https://api.spotify.com/v1/audio-analysis/2V28RKTOabMKk2xvIDyCEz    1
https://api.spotify.com/v1/audio-analysis/6MAAMZImxcvYhRnxDLTufD    1
Name: analysis_url, Length: 35877, dtype: int64

**duration_ms**

In [ ]:
data["duration_ms"].value_counts(dropna=False).sort_index()

25600     1
35862     1
38333     1
41290     1
42133     1
         ..
821168    1
847302    1
855502    1
894386    1
913052    1
Name: duration_ms, Length: 26261, dtype: int64

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["duration_ms"])
# plt.show()

**time_signature**

In [ ]:
data["time_signature"].value_counts(dropna=False).sort_index()

1      150
3     1219
4    40427
5      509
Name: time_signature, dtype: int64

In [ ]:
# fig, axis = plt.subplots(figsize=(5,5))
# axis.hist(data["time_signature"])
# plt.show()

**Genre**

In [ ]:
data["genre"].value_counts(dropna=False).sort_index()

Dark Trap          4578
Emo                1680
Hiphop             3028
Pop                 461
Rap                1848
RnB                2099
Trap Metal         1956
Underground Rap    5875
dnb                2966
hardstyle          2936
psytrance          2961
techhouse          2975
techno             2956
trance             2999
trap               2987
Name: genre, dtype: int64

**Song_name**

In [ ]:
data["song_name"].value_counts().sort_index()

"Enough."                           1
"Life"                              2
"Quarter Century Freestyle"         1
"The Take Over, The Breaks Over"    3
"You said you love me..."           1
                                   ..
• Tales                             1
ムスティネ                               1
泳 (Yong)                            1
ꞌCosmicꞌ.m4a                        2
행복                                  1
Name: song_name, Length: 15439, dtype: int64

In [ ]:
data["song_name"].value_counts(ascending=False,dropna=False)

NaN               20786
Forever              14
Lost                 12
Alone                11
Revenge              11
                  ...  
Salary                1
Patek Philippe        1
ForceQuit             1
Red Nose              1
Can't Hold On         1
Name: song_name, Length: 15440, dtype: int64

**Unnamed: 0**

In [ ]:
data["Unnamed: 0"].value_counts(ascending=False,dropna=False)

NaN        21525
13955.0        1
13973.0        1
13972.0        1
13971.0        1
           ...  
6994.0         1
6993.0         1
6992.0         1
6991.0         1
20999.0        1
Name: Unnamed: 0, Length: 20781, dtype: int64

**Title**

In [ ]:
data["title"].value_counts(ascending=False,dropna=False)

NaN                              21525
Euphoric Hardstyle                1398
UKF Drum & Bass - All Uploads      907
Top Trance Songs EVER              771
I'm a fucking Headbanger           715
                                 ...  
Trance List | TOP Trance            18
EDC 2020 Bass Pod                   16
Best New Techno - When We Dip       13
Trap & Bass 2020                    12
HALCYON's 808 BASEMENT              12
Name: title, Length: 133, dtype: int64

**Duplicates**

In [ ]:
# data[["analysis_url","track_href","uri","id"]].value_counts() #35877, matches with before

In [ ]:
# data[data["id"]=="2KVv2zAjDHxymySY5EIaAz"]

In [ ]:
# # data.loc[data.id=="7D5qMGajdfznFcXFArk5VE", (data.columns != 'Unnamed: 0') & (data.columns != 'title')].duplicated()
# sum(data.loc[:, (data.columns != 'Unnamed: 0') & (data.columns != 'title')].duplicated()) #3218

In [ ]:
numeric_columns = ["time_signature", "duration_ms", "danceability", "energy", "key", "loudness",
                  "mode", "speechiness", "acousticness", "instruentalness", "liveness", "valence", "tempo"]
top500 = data[data.columns[data.columns.isin(numeric_columns)]].head(500)
# sns.pairplot(top500)

Numeric Columns (most useful for analysis):
* Time Signature - An estimated time signature. The time signature (meter) is a notational convention to specify how many beats are in each bar (or measure). The time signature ranges from 3 to 7 indicating time signatures of "3/4", to "7/4".
* duration_ms - duration in millisecond
* danceability - how suitable it is for dancing
* energy - Energy is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.
* key - key track is in. none = -1. maps to pitches
* loudness - The overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude). Values typically range between -60 and 0 db.
* mode - Mode indicates the modality (major or minor) of a track, the type of scale from which its melodic content is derived. Major is represented by 1 and minor is 0.
* speechiness - Speechiness detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.
* acousticness - confidence measure from 0-1 in whether a track is acoustic.
* instrumentalness - predicts whether it has vocals or not
* liveness - Detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.
* valence A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).
* tempo - The overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.

Character Columns (for bar plots and labels):
* Genre

Columns with large amount of NA's:
* title (21525)
* Unnamed: 0 (21525)
* song_name (20786)

Index/Dupe columns:
* song_name (20786 NA's)
* analysis_url (no NAs, but has duplicates)
* track_href (no NAs, but has duplicates, same as analysis_url?)
* uri(no NAs, but has duplicates, same as analysis_url?)
* id (no NAs, but has duplicates, same as analysis_url?) *Use this for dedupe and set as index*; this is Spotify's ID

Useless Columns:
* Type

## Data Cleaning

* **Dedupe.** Use ID as the index and dedupe based on ID.
* **Pivot.** Pivot the title column?
* **Collinearity.** Energy is collinear with loudness

In [ ]:
nid = ["time_signature", "duration_ms", "danceability", "energy", "key", "loudness",
       "mode", "speechiness", "acousticness", "instruentalness", "liveness", "valence", "tempo","id"]
ngid = ["time_signature", "duration_ms", "danceability", "energy", "key", "loudness",
       "mode", "speechiness", "acousticness", "instruentalness", "liveness", "valence", "tempo","id",
       "genre"]
nid
deduped = data.drop_duplicates(subset="id") #35877
deduped_number = deduped[deduped.columns[deduped.columns.isin(ngid)]]
deduped_number

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0             0.831   0.814    2    -7.364     1       0.4200      0.059800   
1             0.719   0.493    8    -7.230     1       0.0794      0.401000   
2             0.850   0.893    5    -4.783     1       0.0623      0.013800   
3             0.476   0.781    0    -4.710     1       0.1030      0.023700   
4             0.798   0.624    2    -7.668     1       0.2930      0.217000   
...             ...     ...  ...       ...   ...          ...           ...   
42298         0.502   0.991    5    -4.333     0       0.2400      0.005540   
42300         0.528   0.693    4    -5.148     1       0.0304      0.031500   
42302         0.361   0.821    8    -3.102     1       0.0505      0.026000   
42303         0.477   0.921    6    -4.777     0       0.0392      0.000551   
42304         0.529   0.945    9    -5.862     1       0.0615      0.001890   

       liveness  valence    tempo                      id  duration_ms  \
0        0.0556   0.3890  156.985  2Vc6NJ9PW9gD9q343XFRKx       124539   
1        0.1180   0.1240  115.080  7pgJBLVz5VmnL7uGHmRj6p       224427   
2        0.3720   0.0391  218.050  0vSWgAlfpye0WCGeNmuNhy        98821   
3        0.1140   0.1750  186.948  0VSXnJqQkwuH2ei1nOQ1nu       123661   
4        0.1660   0.5910  147.988  4jCeguq9rMTlbMmPHuO7S3       123298   
...         ...      ...      ...                     ...          ...   
42298    0.4700   0.0797  150.029  4PMqy2vJjmv6ydGVqtCqPW       200000   
42300    0.1210   0.3940  150.013  46bXU7Sgj7104ZoXxzz9tM       269208   
42302    0.3850   0.1240  154.935  72DAt9Lbpy9EUS29OzQLob       234823   
42303    0.0575   0.4880  150.042  6HXgExFVuE1c3cq9QjFCcU       323200   
42304    0.4140   0.1340  155.047  6MAAMZImxcvYhRnxDLTufD       162161   

       time_signature      genre  
0                   4  Dark Trap  
1                   4  Dark Trap  
2                   4  Dark Trap  
3                   3  Dark Trap  
4                   4  Dark Trap  
...               ...        ...  
42298               4  hardstyle  
42300               4  hardstyle  
42302               4  hardstyle  
42303               4  hardstyle  
42304               4  hardstyle  

[35877 rows x 14 columns]

In [ ]:
arrayed = np.asarray(deduped_number)

## Model

In [ ]:
np.random.seed(0)
shuffled_indices = np.random.permutation(range(len(arrayed)))
# max(shuffled_indices)
# len(deduped)
# shuffled_indices
shuffled = arrayed[shuffled_indices]

In [ ]:
temp1 = pd.DataFrame(shuffled)
temp1

0      1  2       3  4       5         6       7       8        9   \
0      0.686  0.922  0  -6.455  1  0.0493  0.000376   0.235  0.0837    140.0   
1       0.51   0.97  3  -0.383  0   0.145    0.0238     0.5   0.213  149.293   
2      0.947  0.421  5  -8.864  0   0.166    0.0286   0.116   0.308  200.012   
3      0.815  0.946  2  -4.153  1   0.268    0.0802   0.315   0.864  175.014   
4      0.654  0.177  1 -12.522  1  0.0618     0.572  0.0969   0.132   149.89   
...      ...    ... ..     ... ..     ...       ...     ...     ...      ...   
35872  0.869  0.828  1  -8.472  1  0.0551   0.00633   0.105  0.0887  130.001   
35873   0.51  0.891  8   -4.96  0  0.0411    0.0138  0.0966    0.41  174.994   
35874  0.815  0.997  2  -4.143  1   0.359  0.000761   0.087   0.232  149.982   
35875  0.866  0.828  8  -8.704  1   0.203     0.203   0.123   0.543  133.004   
35876  0.556  0.917  9  -6.106  0  0.0461   0.00689   0.303   0.358  139.968   

                           10      11 12               13  
0      14mLqMYtlM5afsZWHl6Ln9  420529  4        psytrance  
1      1gKxIEPFdjySh8V372ZckL  171200  4             trap  
2      1PlLXlCQThvDTqocy7i5ZR  182410  4  Underground Rap  
3      3VUZLAnE17QSUJAzYxgVb1  111833  4        Dark Trap  
4      7dQeS4ErW1iWvnKdrtKwKr  183022  4              RnB  
...                       ...     ... ..              ...  
35872  6OaoFrK9jC7hPfsLE6xAvH  489231  3           techno  
35873  3iKBtYoamggdtSbDL9FUCa  212914  4              dnb  
35874  5HmxePuM7z5vmIoVDDODb9  185621  4             trap  
35875  53cLYZSnsw7BeJybRHPlox  183333  4           techno  
35876  6JoglNv2Hdyo6wXeZIH0e3  206707  4        Dark Trap  

[35877 rows x 14 columns]

In [ ]:
colnames = {
    0: "danceability",
    1: "energy",
    2: "key",
    3: "loudness",
    4: "mode",
    5: "speechiness",
    6: "acousticness",
    7: "liveness",
    8: "valence",
    9: "tempo",
    10: "id",
    11: "duration_ms",
    12: "time_signature",
    13: "genre"
}

In [ ]:
f80 = int(round(len(temp1)*.8,0))
# X = deduped_number[deduped.columns[deduped.columns.isin(nid)]][:f80]
dtrain_x = temp1[temp1.columns[0:13]][:f80]
dtrain_y = temp1[temp1.columns[[10,13]]][:f80]
dtest_x = temp1[temp1.columns[0:13]][f80:]
dtest_y = temp1[temp1.columns[[10,13]]][f80:]

df_names = [dtrain_x, dtrain_y, dtest_x, dtest_y]

[df.rename(columns=colnames, inplace=True) for df in df_names]

dtrain_x

danceability energy key loudness mode speechiness acousticness liveness  \
0            0.686  0.922   0   -6.455    1      0.0493     0.000376    0.235   
1             0.51   0.97   3   -0.383    0       0.145       0.0238      0.5   
2            0.947  0.421   5   -8.864    0       0.166       0.0286    0.116   
3            0.815  0.946   2   -4.153    1       0.268       0.0802    0.315   
4            0.654  0.177   1  -12.522    1      0.0618        0.572   0.0969   
...            ...    ...  ..      ...  ...         ...          ...      ...   
28697         0.56  0.751   4   -6.022    0      0.0364      0.00168    0.168   
28698         0.63  0.994   1   -0.694    0      0.0658      0.00663    0.337   
28699        0.531  0.569   4  -10.566    0      0.0294        0.191    0.177   
28700        0.762  0.738   5   -5.765    0       0.104        0.457    0.115   
28701        0.418  0.892   8   -4.574    1       0.077     0.000385    0.116   

      valence    tempo                      id  
0      0.0837    140.0  14mLqMYtlM5afsZWHl6Ln9  
1       0.213  149.293  1gKxIEPFdjySh8V372ZckL  
2       0.308  200.012  1PlLXlCQThvDTqocy7i5ZR  
3       0.864  175.014  3VUZLAnE17QSUJAzYxgVb1  
4       0.132   149.89  7dQeS4ErW1iWvnKdrtKwKr  
...       ...      ...                     ...  
28697   0.177  187.938  7GZKme55JpmLq3alBDQm11  
28698   0.668  149.996  6ThIqA5MPO2lyWII2uwLT7  
28699   0.237  125.026  6pSGcCDMsXcdSBaklAUBXm  
28700   0.797  154.992  6a2AjOTaVwpbSJoH13MZVa  
28701   0.221  171.977  2ckzkdSzEe1245ONIb5EJk  

[28702 rows x 11 columns]

In [ ]:
dtrain_x.set_index("id", inplace = True)
dtest_x.set_index("id", inplace = True)
dtrain_y.set_index("id", inplace = True)
dtest_y.set_index("id", inplace = True)

In [ ]:
dtrain_x

danceability energy key loudness mode speechiness  \
id                                                                         
14mLqMYtlM5afsZWHl6Ln9        0.686  0.922   0   -6.455    1      0.0493   
1gKxIEPFdjySh8V372ZckL         0.51   0.97   3   -0.383    0       0.145   
1PlLXlCQThvDTqocy7i5ZR        0.947  0.421   5   -8.864    0       0.166   
3VUZLAnE17QSUJAzYxgVb1        0.815  0.946   2   -4.153    1       0.268   
7dQeS4ErW1iWvnKdrtKwKr        0.654  0.177   1  -12.522    1      0.0618   
...                             ...    ...  ..      ...  ...         ...   
7GZKme55JpmLq3alBDQm11         0.56  0.751   4   -6.022    0      0.0364   
6ThIqA5MPO2lyWII2uwLT7         0.63  0.994   1   -0.694    0      0.0658   
6pSGcCDMsXcdSBaklAUBXm        0.531  0.569   4  -10.566    0      0.0294   
6a2AjOTaVwpbSJoH13MZVa        0.762  0.738   5   -5.765    0       0.104   
2ckzkdSzEe1245ONIb5EJk        0.418  0.892   8   -4.574    1       0.077   

                       acousticness liveness valence    tempo  
id                                                             
14mLqMYtlM5afsZWHl6Ln9     0.000376    0.235  0.0837    140.0  
1gKxIEPFdjySh8V372ZckL       0.0238      0.5   0.213  149.293  
1PlLXlCQThvDTqocy7i5ZR       0.0286    0.116   0.308  200.012  
3VUZLAnE17QSUJAzYxgVb1       0.0802    0.315   0.864  175.014  
7dQeS4ErW1iWvnKdrtKwKr        0.572   0.0969   0.132   149.89  
...                             ...      ...     ...      ...  
7GZKme55JpmLq3alBDQm11      0.00168    0.168   0.177  187.938  
6ThIqA5MPO2lyWII2uwLT7      0.00663    0.337   0.668  149.996  
6pSGcCDMsXcdSBaklAUBXm        0.191    0.177   0.237  125.026  
6a2AjOTaVwpbSJoH13MZVa        0.457    0.115   0.797  154.992  
2ckzkdSzEe1245ONIb5EJk     0.000385    0.116   0.221  171.977  

[28702 rows x 10 columns]

In [ ]:
dtrain_x.columns

Index(['danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness',
       'acousticness', 'liveness', 'valence', 'tempo'],
      dtype='object')

In [ ]:
dtrain_y

genre
id                                     
14mLqMYtlM5afsZWHl6Ln9        psytrance
1gKxIEPFdjySh8V372ZckL             trap
1PlLXlCQThvDTqocy7i5ZR  Underground Rap
3VUZLAnE17QSUJAzYxgVb1        Dark Trap
7dQeS4ErW1iWvnKdrtKwKr              RnB
...                                 ...
7GZKme55JpmLq3alBDQm11              Emo
6ThIqA5MPO2lyWII2uwLT7             trap
6pSGcCDMsXcdSBaklAUBXm        Dark Trap
6a2AjOTaVwpbSJoH13MZVa  Underground Rap
2ckzkdSzEe1245ONIb5EJk              dnb

[28702 rows x 1 columns]

In [ ]:
data["genre"].unique()
genre_dict = {"Dark Trap": 0,
             "Underground Rap": 1,
             "Trap Metal": 2,
             "Emo": 3,
             "Rap": 4,
             "RnB": 5,
             "Pop": 6,
             "Hiphop": 7,
             "techhouse": 8,
             "techno": 9,
             "trance": 10,
             "psytrance": 11,
             "trap": 12,
             "dnb": 13,
             "hardstyle": 14}
dtrain_y_num = [genre_dict[i] for i in dtrain_y["genre"]]
dtrain_y_num
dtest_y_num = [genre_dict[i] for i in dtest_y["genre"]]
dtest_y_num

[7,
 12,
 8,
 1,
 0,
 3,
 13,
 4,
 4,
 2,
 11,
 4,
 2,
 1,
 11,
 11,
 0,
 12,
 12,
 0,
 9,
 7,
 12,
 13,
 11,
 9,
 1,
 12,
 0,
 0,
 0,
 1,
 2,
 1,
 1,
 4,
 1,
 12,
 4,
 1,
 4,
 13,
 11,
 10,
 7,
 7,
 4,
 9,
 4,
 0,
 11,
 1,
 10,
 13,
 4,
 10,
 14,
 13,
 9,
 11,
 0,
 2,
 0,
 2,
 11,
 14,
 10,
 12,
 9,
 14,
 11,
 0,
 7,
 8,
 1,
 1,
 3,
 1,
 11,
 12,
 1,
 0,
 1,
 9,
 9,
 12,
 10,
 9,
 1,
 12,
 1,
 14,
 8,
 9,
 0,
 8,
 12,
 14,
 11,
 5,
 4,
 1,
 7,
 14,
 12,
 12,
 1,
 14,
 3,
 1,
 13,
 8,
 1,
 4,
 5,
 1,
 2,
 1,
 13,
 8,
 10,
 11,
 1,
 10,
 0,
 1,
 10,
 10,
 0,
 14,
 1,
 14,
 14,
 1,
 3,
 1,
 13,
 7,
 4,
 0,
 6,
 6,
 1,
 13,
 10,
 8,
 11,
 12,
 1,
 12,
 4,
 0,
 1,
 13,
 0,
 0,
 0,
 0,
 1,
 4,
 0,
 1,
 1,
 5,
 13,
 8,
 14,
 0,
 1,
 13,
 0,
 3,
 7,
 3,
 5,
 1,
 3,
 0,
 1,
 0,
 0,
 8,
 3,
 3,
 1,
 0,
 5,
 5,
 8,
 6,
 14,
 12,
 8,
 0,
 7,
 13,
 5,
 1,
 7,
 0,
 1,
 13,
 7,
 5,
 13,
 0,
 1,
 0,
 8,
 1,
 0,
 2,
 3,
 4,
 11,
 2,
 5,
 12,
 1,
 2,
 6,
 0,
 14,
 1,
 12,
 9,
 4,
 5,
 11,
 1,
 12,
 14,

In [ ]:
X_train = np.array(dtrain_x).astype('float32')
Y_train = np.array(dtrain_y_num)

X_test = np.array(dtest_x).astype('float32')
Y_test = np.array(dtest_y_num)

In [ ]:
# tf.keras.backend.clear_session()
# tf.random.set_seed(0)
# model = tf.keras.Sequential()

# model.add(tf.keras.layers.Dense(
#     units=15,                     # output dim
#     input_shape=[2],             # input dim
#     use_bias=False,              # we included the bias in X
#     activation='softmax',        # apply a sigmoid to the output
#     kernel_initializer=tf.ones_initializer,  # initialize params to 1
# ))
# optimizer = tf.keras.optimizers.Adam(learning_rate=0.01)
# model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer)

In [ ]:
dtrain_y

genre
id                                     
14mLqMYtlM5afsZWHl6Ln9        psytrance
1gKxIEPFdjySh8V372ZckL             trap
1PlLXlCQThvDTqocy7i5ZR  Underground Rap
3VUZLAnE17QSUJAzYxgVb1        Dark Trap
7dQeS4ErW1iWvnKdrtKwKr              RnB
...                                 ...
7GZKme55JpmLq3alBDQm11              Emo
6ThIqA5MPO2lyWII2uwLT7             trap
6pSGcCDMsXcdSBaklAUBXm        Dark Trap
6a2AjOTaVwpbSJoH13MZVa  Underground Rap
2ckzkdSzEe1245ONIb5EJk              dnb

[28702 rows x 1 columns]

In [ ]:
X_train

array([[6.86000e-01, 9.22000e-01, 0.00000e+00, ..., 2.35000e-01,
        8.37000e-02, 1.40000e+02],
       [5.10000e-01, 9.70000e-01, 3.00000e+00, ..., 5.00000e-01,
        2.13000e-01, 1.49293e+02],
       [9.47000e-01, 4.21000e-01, 5.00000e+00, ..., 1.16000e-01,
        3.08000e-01, 2.00012e+02],
       ...,
       [5.31000e-01, 5.69000e-01, 4.00000e+00, ..., 1.77000e-01,
        2.37000e-01, 1.25026e+02],
       [7.62000e-01, 7.38000e-01, 5.00000e+00, ..., 1.15000e-01,
        7.97000e-01, 1.54992e+02],
       [4.18000e-01, 8.92000e-01, 8.00000e+00, ..., 1.16000e-01,
        2.21000e-01, 1.71977e+02]], dtype=float32)

In [ ]:
# # As above, get predictions for the current model first.
# np.random.seed(12345)
# preds = model.predict(X_train)

# # Do a single gradient update.
# history = model.fit(
#   x = X_train,
#   y = Y_train,
#   epochs=20,
#   batch_size=50,
#   verbose=0)

# # Show the loss (before the update) and the new weights.
# loss = history.history['loss'][0]
# weights = model.layers[0].get_weights()[0].T
# # print('predictions:\n', preds[:6])
# print('loss:', loss)
# # print('W:\n', weights)

# test_preds = model.predict(X_test)
# test_preds_labels = np.argmax(test_preds, axis=1)
# accuracy = np.mean(test_preds_labels == Y_test)
# print(f'accuracy: {accuracy}')

In [ ]:
# xx = model.evaluate(X_test, Y_test)
# xx

In [ ]:
# Y_test

In [ ]:
# test_preds_labels

In [ ]:
# # Create a confusion matrix as a 2D array.
# confusion_matrix = tf.math.confusion_matrix(Y_test, test_preds_labels)

# # Use a heatmap plot to display it.
# ax = sns.heatmap(confusion_matrix, annot=True, fmt='.3g', cmap='Blues',
#                  xticklabels=colnames, yticklabels=colnames, cbar=False)

# # Add axis labels.
# ax.set(xlabel='Predicted Label', ylabel='True Label')
# plt.show()

In [ ]:
# features = ['danceability', 'energy', 'key', 'loudness', 'mode','speechiness','acousticness', 'instrumentalness', 'liveness','valence','tempo','duration_ms']
# def plotting(feature):
#     plt.figure(figsize=(20, 10))
#     plot = sns.catplot(x=feature, y='genre', data=deduped_number, kind='bar')
#     plot.set_ylabels('Genre')
#     plt.show()
    
# for x in features:
#     plotting(x)

In [ ]:
# plt.figure (figsize=(15, 12))
# plt.title('Scatterplot of Loudness & Energy for all Genres', fontsize=16)
# sns.scatterplot(deduped_number['energy'],deduped_number['speechiness'],hue=deduped_number['genre'],x_bins=30);

# Random Forest

https://towardsdatascience.com/random-forest-in-python-24d0893d51c0

In [ ]:
# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
# Instantiate model with 500 decision trees
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)
# Train the model on training data
rf.fit(X_train, Y_train)

RandomForestRegressor(n_estimators=500, random_state=42)

In [ ]:
# Use the forest's predict method on the test data
predictions = rf.predict(X_test)# Calculate the absolute errors
errors = abs(predictions - Y_test)# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2))

Mean Absolute Error: 1.9


In [ ]:
predictions

array([ 6.024, 11.924,  5.834, ...,  8.399,  2.594,  8.648])

In [ ]:
Y_test

array([ 7, 12,  8, ..., 12,  9,  0])

In [ ]:
errors

array([0.976, 0.076, 2.166, ..., 3.601, 6.406, 8.648])

In [ ]:
# accuracy = correct / total
accuracy = np.sum(np.round(predictions,0)==Y_test)/Y_test.shape[0]
accuracy

0.31149825783972124

In [ ]:
# !pip install pydot

In [ ]:
# # Import tools needed for visualization
# from sklearn.tree import export_graphviz
# import pydot
# # Pull out one tree from the forest
# tree = rf.estimators_[5]

# export_graphviz(tree, out_file = 'tree.dot', feature_names = dtrain_x.columns, rounded = True, precision = 1)
# # Use dot file to create a graph
# (graph, ) = pydot.graph_from_dot_file('tree.dot')
# # # Write graph to a png file
# # graph.write_png('tree.png')

In [ ]:
# # Limit depth of tree to 3 levels
# rf_small = RandomForestRegressor(n_estimators=10, max_depth = 3)
# rf_small.fit(train_features, train_labels)# Extract the small tree
# tree_small = rf_small.estimators_[5]# Save the tree as a png image
# export_graphviz(tree_small, out_file = 'small_tree.dot', feature_names = feature_list, rounded = True, precision = 1)(graph, ) = pydot.graph_from_dot_file('small_tree.dot')graph.write_png('small_tree.png');